pkg/eventpersisterops/persister/evaluation_event.go

```diff
@@ -0,0 +1,270 @@
+	aoproto "github.com/bucketeer-io/bucketeer/proto/autoops"
+	eventproto "github.com/bucketeer-io/bucketeer/proto/event/client"
+)
+
+const opsEvalKeyPrefix = "autoops:evaluation"
+
+type evalTargetRule struct {
+	id        string
+	clauseIDs []string
+}
+
+type evalEvtUpdater struct {
+	ctx               context.Context
+	featureClient     featureclient.Client
+	autoOpsClient     aoclient.Client
+	eventCounterCache cachev3.EventCounterCache
+	flightgroup       singleflight.Group
+	logger            *zap.Logger
+}
+
+func NewEvalUserCountUpdater(
+	ctx context.Context,
+	featureClient featureclient.Client,
+	autoOpsClient aoclient.Client,
+	eventCounterCache cachev3.EventCounterCache,
+	logger *zap.Logger,
+) Updater {
+	return &evalEvtUpdater{
+		ctx:               ctx,
+		featureClient:     featureClient,
+		autoOpsClient:     autoOpsClient,
+		eventCounterCache: eventCounterCache,
+		logger:            logger,
+	}
+}
+
+func (u *evalEvtUpdater) UpdateUserCounts(ctx context.Context, evt environmentEventMap) map[string]bool {
+	fails := map[string]bool{}
+	for environmentNamespace, events := range evt {
+		for id, event := range events {
+			switch evt := event.(type) {
+			case *eventproto.EvaluationEvent:
+				retriable, err := u.updateUserCount(ctx, environmentNamespace, evt)
+				if err != nil {
+					if err == ErrNoAutoOpsRules || err == ErrAutoOpsRulesNotFound {
+						handledCounter.WithLabelValues(codeNoLink).Inc()
+						u.logger.Debug(
+							"There is no auto ops rules to link the evaluation event",
+							zap.Error(err),
+							zap.String("eventId", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+						continue
+					}
+					if !retriable {
+						u.logger.Error(
+							"Failed to persister evaluation event for auto ops",
+							zap.Error(err),
+							zap.String("eventId", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+					}
+					fails[id] = retriable
+					continue
+				}
+				handledCounter.WithLabelValues(codeLinked).Inc()
+			default:
+				u.logger.Error(
+					"Unexpected message type while trying to persister an evaluation event",
+					zap.String("eventId", id),
+					zap.String("environmentNamespace", environmentNamespace),
+				)
+				fails[id] = false
+			}
+		}
+	}
+	return fails
+}
+
+func (u *evalEvtUpdater) updateUserCount(
+	ctx context.Context,
+	environmentNamespace string,
+	event *eventproto.EvaluationEvent,
+) (bool, error) {
+	// List all the auto ops rules
+	list, err := u.listAutoOpsRules(ctx, environmentNamespace)
+	if err != nil {
+		handledCounter.WithLabelValues(codeFailedToListAutoOpsRules).Inc()
+		return true, err
+	}
+	if len(list) == 0 {
+		return false, ErrNoAutoOpsRules
+	}
+	// Find the rules
+	rules := u.findOpsRules(event.FeatureId, list)
+	if len(rules) == 0 {
+		return false, ErrAutoOpsRulesNotFound
+	}
+	// Find the event rate clauses by variation ID
+	targetRules := []*evalTargetRule{}
```

Comment:
We can define targetRules as a map. The key will be id and value will be clauseIDs.